File: leetcode/Reverse_Linked_List_II.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (m == n) return head;
        int ct = 1;
        ListNode pre = null;
        ListNode cur = head;
        while (ct < m) {
            pre = cur;
            cur = cur.next;
            ++ct;
        }
        ListNode ml = pre;
        ListNode mr = cur;
        ListNode next = cur.next;
        while (ct < n) {
            pre = cur;
            cur = next;
            next = cur.next;
            cur.next = pre;
            ++ct;
        }
        if (ml == null) {
            ml = cur;
        } else {
            ml.next = cur;
        }
        mr.next = next;
        if (m == 1) return ml;
        return head;
    }
}

File: leetcode/Reverse_Linked_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        ListNode *node = head;
        int pos = 0;
        int mid = (m + n) / 2;
        stack<ListNode *> swapping;
        while (node != NULL) {
            ++pos;
            if (pos >= m && pos <= n) {
                if (pos <= mid) {
                    swapping.push(node);
                }
                // Mind the condition for segment having odd items
                if (pos > mid || (pos == mid && mid - m == n - mid)) {
                    ListNode *paired_node = swapping.top();
                    swapping.pop();
                    int tmp = node->val;
                    node->val = paired_node->val;
                    paired_node->val = tmp;
                }
            }
            node = node->next;
        }
        return head;
    }
};
